cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java

```diff
@@ -0,0 +1,246 @@
+ * for some time, it is confirmed to be a slow broker and the finder will report a {@link SlowBrokers}} anomaly
+ * with broker demotion as self-healing proposal. If the metric anomaly still persists for an extended time, the finder
+ * will eventually report another {@link SlowBrokers}} anomaly with broker removal as self-healing proposal.
+ *
+ * Note if there are too many brokers confirmed as slow broker in the same run, the finder will report the {@link SlowBrokers}
+ * anomaly as unfixable. Because this often indicates some serious issue in the cluster and probably requires administrator's
+ * intervention to decide the right remediation strategy.
+ */
+public class SlowBrokerFinder implements MetricAnomalyFinder<BrokerEntity> {
+  private static final Logger LOG = LoggerFactory.getLogger(SlowBrokerFinder.class);
+  public static final String SELF_HEALING_SLOW_BROKERS_REMOVAL_ENABLED_CONFIG = "self.healing.slow.brokers.removal.enabled";
+  public static final String REMOVE_SLOW_BROKERS_CONFIG = "remove.slow.brokers";
+  // Todo: make following configs configurable.
+  // The percentile threshold used to compare latest metric value against historical value.
+  private static final double HISTORY_METRIC_PERCENTILE_THRESHOLD = 90.0;
+  // The margin used to compare latest metric value against historical value.
+  private static final double HISTORY_METRIC_MARGIN = 3.0;
+  // The percentile threshold used to compare last metric value again peers' latest value.
+  private static final double PEER_METRIC_PERCENTILE_THRESHOLD = 50.0;
+  // The margin used to compare last metric value again peers' latest value.
+  private static final double PEER_METRIC_MARGIN = 5.0;
+  // The score threshold to trigger a demotion for slow broker.
+  private static final int SLOW_BROKER_DEMOTION_SCORE = 5;
+  // The score threshold to trigger a removal for slow broker.
+  private static final int SLOW_BROKER_DECOMMISSION_SCORE = 50;
+  // The maximal ratio of slow brokers in the cluster to trigger self-healing operation.
+  private static double SELF_HEALING_UNFIXABLE_RATIO = 0.1;
+  private KafkaCruiseControl _kafkaCruiseControl;
+  private boolean _slowBrokerRemovalEnabled;
+  private final Map<BrokerEntity, Integer> _brokerSlownessScore;
+  private final Map<BrokerEntity, Long> _detectedSlowBrokers;
+  private final Percentile _percentile;
+
+  public SlowBrokerFinder() {
+    _brokerSlownessScore = new HashMap<>();
+    _detectedSlowBrokers = new HashMap<>();
+    _percentile = new Percentile();
+  }
+
+  private Set<BrokerEntity> detectSlowBrokers(Map<BrokerEntity, ValuesAndExtrapolations> metricsHistoryByBroker,
+                                              Map<BrokerEntity, ValuesAndExtrapolations> currentMetricsByBroker) {
+    // Preprocess raw metrics to get the derived metrics for each broker.
+    Map<BrokerEntity, List<Double>> historyValueByBroker = new HashMap<>();
+    Map<BrokerEntity, Double> currentValueByBroker = new HashMap<>();
```

Comment:
I prefer not for this method, since the extracting logic in this method is pretty customized.  For thing, we have a customized filter to filter candidate broker( i.e. filter out broker current does not service traffic); For another thing, a major chunk in this method is about calculating the special derived metric.